CI `pt` product tests flaky due to HiveServer2 communication failures

`pt (delta-lake)` https://github.com/trinodb/trino/actions/runs/32381404383/job/96468931793?pr=30836

```

compatibility (CompatibilityEnvironment):
[ERROR] testSelectOnViewFromDifferentSchema(CompatibilityEnvironment) (CompatibilityEnvironment)
java.lang.RuntimeException: Failed to execute Hive update: DROP DATABASE IF EXISTS test_schema CASCADE
	at io.trino.tests.product.compatibility.CompatibilityEnvironment.executeHiveUpdate(CompatibilityEnvironment.java:140)
	at io.trino.tests.product.hive.TestHiveViewCompatibility.testSelectOnViewFromDifferentSchema(TestHiveViewCompatibility.java:84)
Caused by: java.sql.SQLException: Could not open connection to jdbc:hive2://localhost:32813/default: java.net.SocketException: Broken pipe
	at org.apache.hive.jdbc.HiveConnection.openTransport(HiveConnection.java:206)
	at org.apache.hive.jdbc.HiveConnection.<init>(HiveConnection.java:178)
	at org.apache.hive.jdbc.HiveDriver.connect(HiveDriver.java:105)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:613)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:199)
	at io.trino.tests.product.compatibility.CompatibilityEnvironment.executeHiveUpdate(CompatibilityEnvironment.java:135)
	... 1 more
Caused by: io.trino.hive.jdbc.$internal.org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe
	at io.trino.hive.jdbc.$internal.org.apache.thrift.transport.TIOStreamTransport.flush(TIOStreamTransport.java:161)
	at io.trino.hive.jdbc.$internal.org.apache.thrift.transport.TSaslTransport.sendSaslMessage(TSaslTransport.java:166)
	at io.trino.hive.jdbc.$internal.org.apache.thrift.transport.TSaslClientTransport.handleSaslStartMessage(TSaslClientTransport.java:109)
	at io.trino.hive.jdbc.$internal.org.apache.thrift.transport.TSaslTransport.open(TSaslTransport.java:271)
	at io.trino.hive.jdbc.$internal.org.apache.thrift.transport.TSaslClientTransport.open(TSaslClientTransport.java:38)
	at org.apache.hive.jdbc.HiveConnection.openTransport(HiveConnection.java:203)
	... 6 more
Caused by: java.net.SocketException: Broken pipe
	at java.base/sun.nio.ch.SocketDispatcher.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:65)
	at java.base/sun.nio.ch.NioSocketImpl.tryWrite(NioSocketImpl.java:402)
	at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:418)
	at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:448)
	at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:821)
	at java.base/java.net.Socket$SocketOutputStream.implWrite(Socket.java:1086)
	at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1076)
	at java.base/java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:123)
	at java.base/java.io.BufferedOutputStream.flush(BufferedOutputStream.java:203)
	at io.trino.hive.jdbc.$internal.org.apache.thrift.transport.TIOStreamTransport.flush(TIOStreamTransport.java:159)
	... 11 more

```


this shoule be retrying

--- a/testing/trino-product-tests/src/test/java/io/trino/tests/product/compatibility/CompatibilityEnvironment.java
+++ b/testing/trino-product-tests/src/test/java/io/trino/tests/product/compatibility/CompatibilityEnvironment.java
@@ -31,6 +31,7 @@ import java.sql.Statement;
 import java.time.Duration;
 import java.util.Map;
 
+import static io.trino.testing.containers.environment.ConnectionRetry.createWithRetry;
 import static io.trino.tests.product.hive.HiveCatalogPropertiesBuilder.hadoopMetastoreUri;
 import static java.lang.Integer.parseInt;
 import static java.lang.String.format;
@@ -132,7 +133,7 @@ public class CompatibilityEnvironment
                 "jdbc:hive2://%s:%s/default",
                 hadoop.getHost(),
                 hadoop.getMappedPort(HadoopContainer.HIVESERVER2_PORT));
-        try (Connection connection = DriverManager.getConnection(jdbcUrl, "hive", "");
+        try (Connection connection = createWithRetry(() -> DriverManager.getConnection(jdbcUrl, "hive", ""));
                 Statement statement = connection.createStatement()) {
             return statement.executeUpdate(sql);
         }

--- a/testing/trino-testing-containers/src/main/java/io/trino/testing/containers/environment/ConnectionRetry.java
+++ b/testing/trino-testing-containers/src/main/java/io/trino/testing/containers/environment/ConnectionRetry.java
@@ -1,0 +1,62 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.testing.containers.environment;
+
+import com.google.common.base.Throwables;
+import dev.failsafe.Failsafe;
+import dev.failsafe.FailsafeException;
+import dev.failsafe.RetryPolicy;
+import io.airlift.log.Logger;
+
+import java.net.SocketException;
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import static java.time.temporal.ChronoUnit.SECONDS;
+
+/// Creates JDBC connections with retries for transient socket failures.
+public final class ConnectionRetry
+{
+    private static final Logger log = Logger.get(ConnectionRetry.class);
+
+    private static final RetryPolicy<Object> RETRY_POLICY = RetryPolicy.builder()
+            .handleIf(failure -> Throwables.getRootCause(failure) instanceof SocketException)
+            .withBackoff(1, 10, SECONDS)
+            .withMaxRetries(30)
+            .onRetry(event -> log.warn(event.getLastException(), "Connection failed on attempt %d, will retry.", event.getAttemptCount()))
+            .build();
+
+    private ConnectionRetry() {}
+
+    public static Connection createWithRetry(ConnectionFactory connectionFactory)
+            throws SQLException
+    {
+        try {
+            return Failsafe.with(RETRY_POLICY).get(connectionFactory::create);
+        }
+        catch (FailsafeException failure) {
+            if (failure.getCause() instanceof SQLException sqlException) {
+                throw sqlException;
+            }
+            throw failure;
+        }
+    }
+
+    @FunctionalInterface
+    public interface ConnectionFactory
+    {
+        Connection create()
+                throws SQLException;
+    }
+}

